### DISCUSSION TITLE
Better Understanding Caching

### DISCUSSION THREAD

**Discussion by daymiani** [ranaroussi/yfinance#2151] [General]
My knowledge of caching is very high-level and I'd like to better understand how it appears to be saving so much time in my routine requests for hour/day/week/month S&P data. As suggested in the [docs](https://ranaroussi.github.io/yfinance/advanced/caching.html), I'm using a CachedLimiterSession:

```
session = CachedLimiterSession(
        limiter=Limiter(RequestRate(6, Duration.SECOND)),
        bucket_class=MemoryQueueBucket,
        backend=SQLiteCache("yfinance.cache"),
)

prices['h'] = yf.download(tickers, period='5d', interval='1h', session=session)
prices['d'] = yf.download(tickers, period='5d', interval='1d', session=session)
prices['w'] = yf.download(tickers, period='3mo', interval='1wk', session=session)
prices['m'] = yf.download(tickers, period='6mo', interval='1mo', session=session)

# "tickers" is all stocks in the S&P 500 
```

This job runs every 10 minutes during market hours. On first-run (after a fresh deploy), the above code takes roughly 8 minutes. Subsequent runs however, take only 30 seconds! How is this possible given I need to pull 500 stocks across 4 time intervals (i.e 2000 requests) with a 6 request per second rate limit? Assuming the historical data is cached, the current hour/day/week/month close data (i.e. current price) is always in flux, so isn't a minimum of 2000 requests still needed every time? How does caching achieve such dramatic time savings in my scenario? Thank you kindly for your insight.

**Comment by ValueRaider** [comment]
Start here https://en.wikipedia.org/wiki/Cache_(computing)

**Reply by daymiani** [reply]
Thank you ValueRaider. Unfortunately, I don't see anything in the article which helps address this question:
 
The current close price is always in flux during market hours, so at an absolute minimum (assuming all other data is cached) the current close price for all 500 stocks will need to be pulled. With a six request per second rate limit, wouldn't that require at least 83 seconds? (500/6)

**Reply by ValueRaider** [reply]
request_cache doesn't care about any of that, it just stores GET/POST requests and responses.

**Reply by daymiani** [reply]
Got it, thank you! I erroneously believed requests_cache was doing more than it actually does.

**Reply by ValueRaider** [reply]
https://github.com/ValueRaider/yfinance-cache

**Comment by eduardo-barros** [comment]
What I understand about yfinance.cache:

-  it is a SQLite3 database with 2 tables, "responses" and "redirects"
- "responses" has 3 columns: "key", "value", and "expires".
- "value" is a bytes object that stores an HTTP response with key/value pairs.
- to parse "value", use pickle.loads() - it will return the response as a dict.
- the market data itself is a bytes object under the "_content" key. Use json.loads() to read it.
